ResearchKit/ActiveTasks/ORKAudioGenerator.m

```diff
@@ -0,0 +1,225 @@
+    Float32 *bufferNonActive = (Float32 *)ioData->mBuffers[1 - audioGenerator->activeChannel].mData;
+
+    // Generate the samples
+    for (UInt32 frame = 0; frame < inNumberFrames; frame++) {
+        double bufferValue = sin(theta) * amplitude * fadeInFactor;
+        bufferActive[frame] = bufferValue;
+        bufferNonActive[frame] = 0;
+
+        // printf("%f, %f\n", bufferValue, 20*log(bufferValue));
+
+        theta += theta_increment;
+        if (theta > 2.0 * M_PI) {
+            theta -= 2.0 * M_PI;
+        }
+
+        fadeInFactor += 1/(ORKSineWaveToneGeneratorSampleRateDefault * audioGenerator->fadeInDuration);
+        if (fadeInFactor >= 1) {
+            fadeInFactor = 1;
+        }
+    }
+
+    // Store the theta back in the view controller
+    audioGenerator->theta = theta;
+    audioGenerator->fadeInFactor = fadeInFactor;
+
+    return noErr;
+}
+
+
+@implementation ORKAudioGenerator
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        [self setupAudioSession];
+    }
+    return self;
+}
+
+- (double)volumeInDecibels {
+    return 20 * log(ORKSineWaveToneGeneratorAmplitudeDefault * fadeInFactor);
+}
+
+- (double)volumeAmplitude {
+    return ORKSineWaveToneGeneratorAmplitudeDefault * fadeInFactor;
+}
+
+- (void)playSoundAtFrequency:(double)playFrequency
+                   onChannel:(ORKAudioChannel)playChannel
+              fadeInDuration:(NSTimeInterval)duration {
+    frequency = playFrequency;
+    activeChannel = playChannel;
+    fadeInFactor = 0;
+    fadeInDuration = duration;
+
+    [self play];
+}
+
+- (void)play {
+    if (!_toneUnit) {
+        [self createToneUnit];
+
+        // Stop changing parameters on the unit
+        OSErr err = AudioUnitInitialize(_toneUnit);
+        NSAssert1(err == noErr, @"Error initializing unit: %hd", err);
+
+        // Start playback
+        err = AudioOutputUnitStart(_toneUnit);
+        NSAssert1(err == noErr, @"Error starting unit: %hd", err);
+    }
+}
+
+- (void)stop {
+    if (_toneUnit) {
+        AudioOutputUnitStop(_toneUnit);
+        AudioUnitUninitialize(_toneUnit);
+        AudioComponentInstanceDispose(_toneUnit);
+        _toneUnit = nil;
+    }
+}
+
+- (void)setupAudioSession {
+    AVAudioSession *audioSession = [AVAudioSession sharedInstance];
+    BOOL ok;
+    NSError *setCategoryError = nil;
+    ok = [audioSession setCategory:AVAudioSessionCategoryPlayback error:&setCategoryError];
+    NSAssert1(ok, @"Audio error %@", setCategoryError);
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(handleInterruption:)
+                                                 name:AVAudioSessionInterruptionNotification
+                                               object:audioSession];
+}
+
+- (void)createToneUnit {
+    // Configure the search parameters to find the default playback output unit
```

Comment:
What's the original source of this code? I've found this exact comment all over the web, so this code has probably come from an example.